builder/common_builder/common_builder.go

```diff
@@ -0,0 +1,134 @@
+	"github.com/ercole-io/ercole-agent/fetcher"
+	"github.com/ercole-io/ercole-agent/logger"
+	"github.com/ercole-io/ercole-agent/model"
+	"github.com/ercole-io/ercole-agent/utils"
+)
+
+// CommonBuilder for Linux and Windows hosts
+type CommonBuilder struct {
+	fetcher       fetcher.Fetcher
+	configuration config.Configuration
+	log           logger.Logger
+}
+
+// NewCommonBuilder initialize an appropriate builder for Linux or Windows
+func NewCommonBuilder(configuration config.Configuration, log logger.Logger) CommonBuilder {
+	var specializedFetcher fetcher.SpecializedFetcher
+
+	log.Debugf("runtime.GOOS: [%v]", runtime.GOOS)
+
+	if runtime.GOOS == "windows" {
+		wf := fetcher.NewWindowsFetcherImpl(configuration, log)
+		specializedFetcher = &wf
+
+	} else {
+		if runtime.GOOS != "linux" {
+			log.Errorf("Unknow runtime.GOOS: [%v], I'll try with linux\n", runtime.GOOS)
+		}
+
+		lf := fetcher.NewLinuxFetcherImpl(configuration, log)
+		specializedFetcher = &lf
+	}
+
+	fetcherImpl := &fetcher.CommonFetcherImpl{
+		SpecializedFetcher: specializedFetcher,
+		Configuration:      configuration,
+		Log:                log,
+	}
+
+	builder := CommonBuilder{
+		fetcher:       fetcherImpl,
+		configuration: configuration,
+		log:           log,
+	}
+
+	return builder
+}
+
+// Run fill hostData
+func (b *CommonBuilder) Run(hostData *model.HostData) {
+	if b.configuration.Features.Exadata.Enabled {
+		b.log.Debug("Exadata mode enabled")
+
+		b.checksToRunExadata()
+
+		if err := b.fetcher.SetUser(b.configuration.Features.Exadata.NotRootUser); err != nil {
+			b.log.Panicf("Can't set user [%s] for fetcher, err: [%v]", b.configuration.Features.Exadata.NotRootUser, err)
+		}
+	}
+
+	hostData.Info = *b.getHost()
+
+	hostData.Hostname = hostData.Info.Hostname
+	if b.configuration.Hostname != "default" {
+		hostData.Hostname = b.configuration.Hostname
+	}
+
+	if b.configuration.Features.Databases.Enabled {
+		b.log.Debug("Databases mode enabled")
+		hostData.Extra.Filesystems = b.fetcher.GetFilesystems()
+
+		hostData.Extra.Databases = b.getOracleDBs(hostData.Info.Type)
+		hostData.Databases, hostData.Schemas = b.getDatabasesAndSchemaNames(hostData.Extra.Databases)
+	}
+
+	if b.configuration.Features.Virtualization.Enabled {
+		b.log.Debug("Virtualization mode enabled")
+		hostData.Extra.Clusters = b.getClustersInfos()
+	}
+
+	if b.configuration.Features.Exadata.Enabled {
+		b.log.Debug("Exadata mode enabled")
+
+		if err := b.fetcher.SetUserAsCurrent(); err != nil {
+			b.log.Panicf("Can't set current user for fetcher, err: [%v]", err)
+		}
+
+		hostData.Extra.ExadataDevices = b.getExadataDevices()
+	}
+}
+
+func (b *CommonBuilder) checksToRunExadata() {
+	if runtime.GOOS != "linux" {
+		b.log.Panicf("Can't run exadata mode if os is different from linux, current os: [%v]", runtime.GOOS)
+	}
+
+	if strings.TrimSpace(b.configuration.Features.Exadata.NotRootUser) == "" {
+		b.log.Panicf("You must set NotRootUser to run in exadata mode")
+	}
```

Comment:
Why NotRootUser is a requirement? Exadata fetchers require root permission?

Reply:
Exadata functionalities must run as root while other "vanilla" ercole-agent functionalites as NotRootUser user, as stated by issue.
So if NotRootUser isn't set I don't know what to do: I ask @raffaelloferrari ?!

Reply:
If NotRootUser is not setted you should run the fetchers with the current user.

Reply:
Done :+1: